mypy/build.py

```diff
@@ -861,59 +863,136 @@ def is_file(path: str) -> bool:
+                                   stderr=subprocess.PIPE).decode()
+
+
+@functools.lru_cache(maxsize=None)
+def get_site_packages_dirs(python_executable: str) -> List[str]:
+    """Find package directories for given python."""
+    if python_executable == sys.executable:
+        # Use running Python's package dirs
+        if hasattr(site, 'getusersitepackages') and hasattr(site, 'getsitepackages'):
+            user_dir = site.getusersitepackages()
+            return site.getsitepackages() + [user_dir]
+        # If site doesn't have get(user)sitepackages, we are running in a
+        # virtualenv, and should fall back to get_python_lib
+        return [get_python_lib()]
+    else:
+        # Use subprocess to get the package directory of given Python
+        # executable
+        try:
+            output = call_python(python_executable, USER_SITE_PACKAGES)
+        except subprocess.CalledProcessError:
+            # if no paths are found (raising a CalledProcessError), we fall back on sysconfig,
+            # the python executable is likely in a virtual environment, thus lacking
+            # needed site methods
+            output = call_python(python_executable, VIRTUALENV_SITE_PACKAGES)
+        return [line for line in output.splitlines() if os.path.isdir(line)]
+
+
+def find_base_dirs(lib_path: Tuple[str, ...], dir_chain: str, components: List[str],
+                   site_packages_dirs: List[str]) -> Tuple[Iterable[str], List[str]]:
+    # If we're looking for a module like 'foo.bar.baz', it's likely that most of the
+    # many elements of lib_path don't even have a subdirectory 'foo/bar'.  Discover
+    # that only once and cache it for when we look for modules like 'foo.bar.blah'
+    # that will require the same subdirectory.
+    dirs = find_module_dir_cache.get((dir_chain, lib_path), [])
+    if not dirs:
+        # Regular packages on the PATH
+        for pathitem in lib_path:
+            # e.g., '/usr/lib/python3.4/foo/bar'
+            isdir = find_module_isdir_cache.get((pathitem, dir_chain))
+            if isdir is None:
+                dir = os.path.normpath(os.path.join(pathitem, dir_chain))
+                isdir = os.path.isdir(dir)
+                find_module_isdir_cache[pathitem, dir_chain] = isdir
+            if isdir:
+                dirs.append(dir)
+        find_module_dir_cache[dir_chain, lib_path] = dirs
+    third_party_dirs = []
+    # Third-party stub/typed packages
+    for pkg_dir in site_packages_dirs:
+        prefix = components[0]
+        stub_name = components[0] + '-stubs'
+        typed_file = os.path.join(pkg_dir, components[0], 'py.typed')
+        stub_dir = os.path.join(pkg_dir, stub_name)
+        if os.path.isdir(stub_dir):
+            components[0] = stub_name
+            rest = components[:-1]
+            path = os.path.join(pkg_dir, *rest)
+            if os.path.isdir(path):
+                third_party_dirs.append(path)
+            components[0] = prefix
```

Comment:
Why do you modify This would be clearer as:
```
stub_components = [stub_name] + components[1:]
path = os.path.join(pkg_dir, *stub_components[:-1])
```
Rather than mutating the `components` list